hal/src/thumbv7em/clock/v2/xosc.rs

```diff
@@ -26,169 +26,125 @@ use typenum::U0;
-    impl XoscNum for Xosc0 {
-        const NUM: usize = 0;
-        const DPLL_SRC: DpllSourceEnum = DpllSourceEnum::XOSC0;
-        type XIn = PA14;
-        type XOut = PA15;
-    }
-
-    impl GclkSourceMarker for Xosc0 {
-        const GCLK_SRC: GclkSourceEnum = GclkSourceEnum::XOSC0;
-    }
-
-    impl DpllSourceMarker for Xosc0 {
-        const DPLL_SRC: DpllSourceEnum = DpllSourceEnum::XOSC0;
-    }
-
-    impl SourceMarker for Xosc0 {}
-
-    impl NotGclkInput for Xosc0 {}
-
-    /// Type which serves as a source marker for the [`super::Xosc1`] and
-    /// provides numerical identity for it
-    pub enum Xosc1 {}
-
-    impl Sealed for Xosc1 {}
-
-    impl XoscNum for Xosc1 {
-        const NUM: usize = 1;
-        const DPLL_SRC: DpllSourceEnum = DpllSourceEnum::XOSC1;
-        type XIn = PB22;
-        type XOut = PB23;
-    }
-
-    impl GclkSourceMarker for Xosc1 {
-        const GCLK_SRC: GclkSourceEnum = GclkSourceEnum::XOSC1;
-    }
-
-    impl DpllSourceMarker for Xosc1 {
-        const DPLL_SRC: DpllSourceEnum = DpllSourceEnum::XOSC1;
-    }
+/// Type-level variant representing the identity of XOSC0
+///
+/// This type is a member of several [type-level enums]. See the documentation
+/// on [type-level enums] for more details on the pattern.
+///
+/// [type-level enums]: crate::typelevel#type-level-enum
+pub enum XoscId0 {}
 
-    impl SourceMarker for Xosc1 {}
+impl Sealed for XoscId0 {}
 
-    impl NotGclkInput for Xosc1 {}
+/// Type which serves as a source marker for the [`super::Xosc0`] and
+/// provides numerical identity for it
+impl XoscId for XoscId0 {
+    const NUM: usize = 0;
+    type XIn = PA14;
+    type XOut = PA15;
 }
 
-/// Encapsulates a hardware crystal configuration
+/// Type-level variant representing the identity of XOSC1
 ///
-/// Supported crystal frequencies: between 8 MHz and 48 MHz
+/// This type is a member of several [type-level enums]. See the documentation
+/// on [type-level enums] for more details on the pattern.
 ///
-/// User must provide the correct hardware frequency here,
-/// that allows configuring hardware to optimise for specific frequencies used.
-pub struct CrystalConfig {
-    src_freq: Hertz,
-    current: CrystalCurrent,
-}
+/// [type-level enums]: crate::typelevel#type-level-enum
+pub enum XoscId1 {}
 
-impl CrystalConfig {
-    /// Create a new [`CrystalConfig`] running at frequency `src_freq`
-    pub fn new(src_freq: impl Into<Hertz>) -> Result<Self, ()> {
-        // Calculate the CrystalCurrent from the supplied
-        // crystal frequency
-        let (current, src_freq) = match src_freq.into().0 {
-            freq @ 8_000_000 => (CrystalCurrent::BaseFreq8m, freq),
-            freq @ 8_000_001..=16_000_000 => (CrystalCurrent::LowFreq8mTo16m, freq),
-            freq @ 16_000_001..=24_000_000 => (CrystalCurrent::MedFreq16mTo24m, freq),
-            freq @ 24_000_001..=48_000_000 => (CrystalCurrent::HighFreq24mTo48m, freq),
-            _ => return Err(()),
```

Comment:
Huh? The [existing implementation](https://github.com/vccggorski/atsamd/blob/a7378f5f5a28c84c028a7e1f5fca2c5f8295b454/hal/src/thumbv7em/clock/v2/xosc.rs#L131-L140) converts the `src_freq` into a `CrystalCurrent` inside the `new` function, returning an `Err(())` if it's out of bounds.

I agree that `TryFrom<Hertz>` or some other fallible API is better in the general case. But in this specific case, I don't think that makes sense. Consider the reasons the it might be out of bounds to begin with:
- The frequency is statically known and valid, but  the user made a typo.
    - I think `panic!` is the best choice here.
- The frequency is statically known and truly outside the 8-48 MHz range.
    - I don't think this is supported by the XOSC. Again, `panic!` makes sense.
- The frequency is not statically known and can be changed.
    - I suppose this is possible, but I've never seen or heard of a dynamically adjustable crystal oscillator. Have you?

I guess my questions is this: What can a user *do* with an `Err` other than `.unwrap()` it? If every reasonable path leads to `.unwrap()`, then why bother returning a `Result` in the first place?

Reply:
One other thought. Maybe we should just force the user to specify *both* the frequency and current. That way there's no possibility of error or panic. And we can repeat the datasheet documentation on the CrystalCurrent struct, so users don't have to go digging in the datasheet. Do you like that better?

Reply:
The more I think about it, the more I like forcing the user to make a choice. We can auto-select a `CrystalCurrent` based on the frequency, but there are quite a few situations where the choice is not optimal or undefined. In those situations, I think it makes sense to push the decision to the user.

The `from_crystal` function would take both a frequency and a `CrystalCurrent`. Thoughts?

Reply:
I implemented this change. Now, the user is forced to manually choose a current in `from_crystal`. The documentation for `CrystalCurrent` provides a summary of the datasheet. Does that work?

Reply:
I support this idea that the user must make a conscious decision for both frequency and current.
As noted they are not set in stone (frequency - current relation) and with the datasheet documentation provided through docs users easily can find what to pick.